src/backend/optimizer/path/allpaths.c

```diff
@@ -1700,9 +1700,12 @@ set_subquery_pathlist(PlannerInfo *root, RelOptInfo *rel,
 		 * we can only gather it to singleQE and materialize the data because we
 		 * cannot pass params across motion.
 		 */
-		if ((!bms_is_empty(required_outer)) &&
-			check_query_for_possible_motion(subquery))
-			config->force_singleQE = true;
+		if (!bms_is_empty(required_outer))
+		{
+			config->is_under_subplan = true;
+			if (check_query_for_possible_motion(subquery))
+				config->force_singleQE = true;
+		}
```

Comment:
why not
```c
		if ((!bms_is_empty(required_outer)) &&
			check_query_for_possible_motion(subquery))
		{
			config->force_singleQE = true;
			config->is_under_subplan = true;
		}
```
or not
```c
		config->is_under_subplan = true;
		if ((!bms_is_empty(required_outer)) &&
			check_query_for_possible_motion(subquery))
			config->force_singleQE = true;
```
?

Reply:
Could you then add tests for these two cases to make sure they don't break?

Reply:
It seems like `check_query_for_possible_motion` activates simply by having an aggregate function in the subquery, so your first suggestion would also work. Should I change it to that? I will also add a test for regular joins still being optimized correctly.

Reply:
Yes, I think we should have three cases in our tests.